oak-search/src/main/java/org/apache/jackrabbit/oak/plugins/index/search/spi/editor/FulltextIndexEditor.java

```diff
@@ -285,61 +296,73 @@ public void markDirty() {
     }
 
     private MatcherState getMatcherState(String name, NodeState after) {
-        List<Aggregate.Matcher> matched = new ArrayList<>();
-        List<Aggregate.Matcher> inherited = new ArrayList<>();
+        List<Aggregate.Matcher> matched = EMPTY_AGGREGATE_MATCHER_LIST;
+        List<Aggregate.Matcher> inherited = EMPTY_AGGREGATE_MATCHER_LIST;
         for (Aggregate.Matcher m : IterableUtils.chainedIterable(matcherState.inherited, currentMatchers)) {
             Aggregate.Matcher result = m.match(name, after);
             if (result.getStatus() == Aggregate.Matcher.Status.MATCH_FOUND) {
+                if (matched == EMPTY_AGGREGATE_MATCHER_LIST) {
+                    matched = new ArrayList<>(4);
+                }
                 matched.add(result);
             }
 
             if (result.getStatus() != Aggregate.Matcher.Status.FAIL) {
+                if (inherited == EMPTY_AGGREGATE_MATCHER_LIST) {
+                    inherited = new ArrayList<>(4);
+                }
                 inherited.addAll(result.nextSet());
             }
         }
 
-        if (!matched.isEmpty() || !inherited.isEmpty()) {
+        if (matched.isEmpty() && inherited.isEmpty()) {
+            return MatcherState.NONE;
+        } else {
             return new MatcherState(matched, inherited);
         }
-        return MatcherState.NONE;
     }
 
 
+
     /*
      * Determines which all matchers are affected by this property change
      *
      * @param name modified property name
      */
     private void checkAggregates(String name) {
-        for (Aggregate.Matcher m : matcherState.matched) {
-            if (!matcherState.affectedMatchers.contains(m)
-                    && m.aggregatesProperty(name)) {
-                matcherState.affectedMatchers.add(m);
+        // Avoid allocating the iterator over matcherState.matched in the common case of the list being empty
+        if (!matcherState.matched.isEmpty()) {
```

Comment:
I first thought there is no measurable performance impact in this additional "if", but then I did measure and found that yes, indeed the additional condition does help. For things that are called millions of times, this should help a bit.

Unfortunately, Java doesn't support macros (custom "for" loops) like [some other languages](https://github.com/thomasmueller/bau-lang?tab=readme-ov-file#macros-and-ternary-condition)...